### PR TITLE
Adding VGG models to the classification task and refactoring

### PR DESCRIPTION
**Pull request by calebrob6**
Pretty much any new class of models we add to the classification task (or soon-to-be segmentation task) will need to have its first conv layer replaced. This PR refactors that logic and adds VGG as a template for "how to add new types of models to the trainers".

### CODE REVIEW

**Comment by calebrob6** [comment]
Pytest on Windows failed with an out of memory error

**Comment by adamjstewart** [comment]
> Pytest on Windows failed with an out of memory error

I've seen this a few times now, not sure if we are just getting unlucky with shared resources or what. We can't do a lot about this since a lot of our models are very memory intensive.

I've also seen a lot of transient matplotlib/tkinter issues on Windows. I think those can be solved by mocking matplotlib to skip the actual plotting step, but I haven't had the time to investigate where that needs to happen yet.

For the time being, you can simply "re-run all tests" to work around these transient issues.

File: torchgeo/trainers/utils.py [hunk 1]
```diff
@@ -95,3 +97,51 @@ def load_state_dict(model: Module, state_dict: Dict[str, Tensor]) -> Module:
     model.load_state_dict(state_dict, strict=False)  # type: ignore[arg-type]
 
     return model
+
+
+def reinit_initial_conv_layer(
```
Comment (adamjstewart) [inline torchgeo/trainers/utils.py:102]: I think there was some talk on Slack about using timm to do this for us? Is that still something we are considering as an alternative to this?

File: torchgeo/trainers/utils.py [hunk 2]
```diff
@@ -95,3 +97,51 @@ def load_state_dict(model: Module, state_dict: Dict[str, Tensor]) -> Module:
...
+    layer: Conv2d, new_in_channels: int, keep_rgb_weights: bool
+) -> Conv2d:
+    """Clones a Conv2d layer while optionally retaining some of the original weights.
+
+    When replacing the first convolutional layer in a model with one that operates over
+    different number of input channels, we sometimes want to keep a subset of the kernel
+    weights the same (e.g. the RGB weights of an ImageNet pretrained model). This is a
+    convinience function that performs that function.
```
Comment (adamjstewart) [inline torchgeo/trainers/utils.py:110]:
```suggestion
    convenience function that performs that function.
```

File: torchgeo/trainers/utils.py [hunk 3]
```diff
@@ -95,3 +97,51 @@ def load_state_dict(model: Module, state_dict: Dict[str, Tensor]) -> Module:
...
+
+    Args:
+        layer: the Conv2d layer to initialize
+        new_in_channels: the new number of input
```
Comment (adamjstewart) [inline torchgeo/trainers/utils.py:114]:
```suggestion
        new_in_channels: the new number of input channels
```

File: torchgeo/trainers/utils.py [hunk 4]
```diff
@@ -95,3 +97,51 @@ def load_state_dict(model: Module, state_dict: Dict[str, Tensor]) -> Module:
...
+        if use_bias:
+            # mypy doesn't realize that bias isn't None here...
+            b_old = layer.bias.data.clone()  # type: ignore[union-attr]
+
+    new_layer = Conv2d(
+        new_in_channels,
+        layer.out_channels,
+        kernel_size=layer.kernel_size,  # type: ignore[arg-type]
+        stride=layer.stride,  # type: ignore[arg-type]
+        padding=layer.padding,  # type: ignore[arg-type]
```
Comment (adamjstewart) [inline torchgeo/trainers/utils.py:132]: For So2Sat we change the stride and padding, right? Maybe these should be parameters of the `reinit_initial_conv_layer` function.

File: torchgeo/trainers/utils.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/trainers/utils.py:102]: Yes, it was kicked off by thinking about this PR

File: torchgeo/trainers/utils.py (hunk 4 above)
Comment (calebrob6) [reply torchgeo/trainers/utils.py:132]: Yeah good point!

**Comment by calebrob6** [comment]
Just rebased

Things to do:
- [ ] Add timm as a dependency
- [ ] Redo `config_model` in `ClassificationTask` to use timm instead of current approach
- [ ] Make sure the tests know about the above changes

**Comment by calebrob6** [comment]
Agree, new PR though?

**Comment by adamjstewart** [comment]
Will review this in more detail after my dentist appt but from a first glance everything looks okay.

**Review by adamjstewart — APPROVED** [review]
Will give @isaaccorley a chance to review before merging

**Comment by isaaccorley** [comment]
This lgtm. I'm excited to try this out.

**Review by isaaccorley — APPROVED** [review]

**Pull request merged**